## PIZZA RUNNER SQL CHALLENGE
This is a sql project created by [Danny Ma]("https://www.linkedin.com/in/datawithdanny/") to help individuals emerging into the data scene. From my view, it helps establish the foundational knowledge of sql while testing and developing logical problem skills.
I have chosen to do it in Python because it highlights how you can create a database on your system to allow for continuous use.

To get access to this and other of Danny's projects - [8 Week SQL Challenge]("https://8weeksqlchallenge.com/").

Before getting started, I would also recommend installing "ipython-sql". This allows you use the 'jupyter magic' function to interact with your relational database.

### Importing Libraries

In [1]:
import sqlite3 as sql
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

### Creating Database, Table & Inserting Values into Table

In [2]:
conn = sql.connect('pizza_runner.db')
c = conn.cursor()

def create_table():
    c.execute('CREATE TABLE IF NOT EXISTS runners (runner_id INT,registration_date TEXT)')
    c.execute('CREATE TABLE IF NOT EXISTS customer_orders (order_id INT,customer_id INT,pizza_id INT,exclusions TEXT,extras TEXT,order_time TEXT)')
    c.execute('CREATE TABLE IF NOT EXISTS runner_orders (order_id INT,runner_id INT,pickup_time TEXT,distance TEXT,duration TEXT,cancellation TEXT)')
    c.execute('CREATE TABLE IF NOT EXISTS pizza_names (pizza_id INT,pizza_name TEXT)')
    c.execute('CREATE TABLE IF NOT EXISTS pizza_recipes (pizza_id INT,toppings TEXT)')
    c.execute('CREATE TABLE IF NOT EXISTS pizza_toppings (topping_id INT,topping_name TEXT)')
    c.execute('CREATE TABLE IF NOT EXISTS ratings (order_id INT,ratings INT,rating_time TEXT)')
    
def data_entry():
    c.execute("INSERT INTO runners VALUES (1, '2021-01-01')")
    c.execute("INSERT INTO runners VALUES (2, '2021-01-03')")
    c.execute("INSERT INTO runners VALUES (3, '2021-01-08')")
    c.execute("INSERT INTO runners VALUES (4, '2021-01-15')")
    
    c.execute("INSERT INTO customer_orders VALUES ('1', '101', '1', '', '', '2020-01-01 18:05:02')")
    c.execute("INSERT INTO customer_orders VALUES ('2', '101', '1', '', '', '2020-01-01 19:00:52')")
    c.execute("INSERT INTO customer_orders VALUES ('3', '102', '1', '', '', '2020-01-02 23:51:23')")
    c.execute("INSERT INTO customer_orders VALUES ('3', '102', '2', '', NULL, '2020-01-02 23:51:23')")
    c.execute("INSERT INTO customer_orders VALUES ('4', '103', '1', '4', '', '2020-01-04 13:23:46')")
    c.execute("INSERT INTO customer_orders VALUES ('4', '103', '1', '4', '', '2020-01-04 13:23:46')")
    c.execute("INSERT INTO customer_orders VALUES ('4', '103', '2', '4', '', '2020-01-04 13:23:46')")
    c.execute("INSERT INTO customer_orders VALUES ('5', '104', '1', 'null', '1', '2020-01-08 21:00:29')")
    c.execute("INSERT INTO customer_orders VALUES ('6', '101', '2', 'null', 'null', '2020-01-08 21:03:13')")
    c.execute("INSERT INTO customer_orders VALUES ('7', '105', '2', 'null', '1', '2020-01-08 21:20:29')")
    c.execute("INSERT INTO customer_orders VALUES ('8', '102', '1', 'null', 'null', '2020-01-09 23:54:33')")
    c.execute("INSERT INTO customer_orders VALUES ('9', '103', '1', '4', '1, 5', '2020-01-10 11:22:59')")
    c.execute("INSERT INTO customer_orders VALUES ('10', '104', '1', 'null', 'null', '2020-01-11 18:34:49')")
    c.execute("INSERT INTO customer_orders VALUES ('10', '104', '1', '2, 6', '1, 4', '2020-01-11 18:34:49')")
    
    c.execute("INSERT INTO runner_orders VALUES ('1', '1', '2020-01-01 18:15:34', '20km', '32 minutes', '')")
    c.execute("INSERT INTO runner_orders VALUES ('2', '1', '2020-01-01 19:10:54', '20km', '27 minutes', '')")
    c.execute("INSERT INTO runner_orders VALUES ('3', '1', '2020-01-03 00:12:37', '13.4km', '20 mins', NULL)")
    c.execute("INSERT INTO runner_orders VALUES ('4', '2', '2020-01-04 13:53:03', '23.4', '40', NULL)")
    c.execute("INSERT INTO runner_orders VALUES ('5', '3', '2020-01-08 21:10:57', '10', '15', NULL)")
    c.execute("INSERT INTO runner_orders VALUES ('6', '3', 'null', 'null', 'null', 'Restaurant Cancellation')")
    c.execute("INSERT INTO runner_orders VALUES ('7', '2', '2020-01-08 21:30:45', '25km', '25mins', 'null')")
    c.execute("INSERT INTO runner_orders VALUES ('8', '2', '2020-01-10 00:15:02', '23.4 km', '15 minute', 'null')")
    c.execute("INSERT INTO runner_orders VALUES ('9', '2', 'null', 'null', 'null', 'Customer Cancellation')")
    c.execute("INSERT INTO runner_orders VALUES ('10', '1', '2020-01-11 18:50:20', '10km', '10minutes', 'null')")
              
    c.execute("INSERT INTO pizza_names VALUES (1, 'Meat Lovers')")
    c.execute("INSERT INTO pizza_names VALUES (2, 'Vegetarian')")
    
    c.execute("INSERT INTO pizza_recipes VALUES (1, '1, 2, 3, 4, 5, 6, 8, 10')")
    c.execute("INSERT INTO pizza_recipes VALUES (2, '4, 6, 7, 9, 11, 12')")
    
    c.execute("INSERT INTO pizza_toppings VALUES (1, 'Bacon')")
    c.execute("INSERT INTO pizza_toppings VALUES (2, 'BBQ Sauce')")
    c.execute("INSERT INTO pizza_toppings VALUES (3, 'Beef')")
    c.execute("INSERT INTO pizza_toppings VALUES (4, 'Cheese')")
    c.execute("INSERT INTO pizza_toppings VALUES (5, 'Chicken')")
    c.execute("INSERT INTO pizza_toppings VALUES (6, 'Mushrooms')")
    c.execute("INSERT INTO pizza_toppings VALUES (7, 'Onions')")
    c.execute("INSERT INTO pizza_toppings VALUES (8, 'Pepperoni')")
    c.execute("INSERT INTO pizza_toppings VALUES (9, 'Peppers')")
    c.execute("INSERT INTO pizza_toppings VALUES (10, 'Salami')")
    c.execute("INSERT INTO pizza_toppings VALUES (11, 'Tomatoes')")
    c.execute("INSERT INTO pizza_toppings VALUES (12, 'Tomato Sauce')")
    
    c.execute("INSERT INTO ratings VALUES ('1', '5', '2020-01-01 17:20:00')")
    c.execute("INSERT INTO ratings VALUES ('2', '5', '2020-01-01 18:30:34')")
    c.execute("INSERT INTO ratings VALUES ('3', '4', '2020-01-03 01:10:47')")
    c.execute("INSERT INTO ratings VALUES ('4', '3', '2020-01-04 15:33:33')")
    c.execute("INSERT INTO ratings VALUES ('5', '3', '2020-01-08 23:20:31')")
    c.execute("INSERT INTO ratings VALUES ('6', 'null', 'null')")
    c.execute("INSERT INTO ratings VALUES ('7', '3', '2020-01-09 20:30:45')")
    c.execute("INSERT INTO ratings VALUES ('8', '4', '2020-01-10 01:15:02')")
    c.execute("INSERT INTO ratings VALUES ('9', 'null', 'null')")
    c.execute("INSERT INTO ratings VALUES ('10', '2', '2020-01-12 12:50:20')")
    
    conn.commit()
    #conn.close()
    
create_table()
data_entry()

#### Loading Database

In [3]:
%%capture
%load_ext sql
%sql sqlite:///pizza_runner.db

#### Checking created tables

In [4]:
%%sql
select 
* 
from runners;

 * sqlite:///pizza_runner.db
Done.


[(1, '2021-01-01'), (2, '2021-01-03'), (3, '2021-01-08'), (4, '2021-01-15')]

In [5]:
%%sql
select 
* 
from runner_orders;

 * sqlite:///pizza_runner.db
Done.


[(1, 1, '2020-01-01 18:15:34', '20km', '32 minutes', ''),
 (2, 1, '2020-01-01 19:10:54', '20km', '27 minutes', ''),
 (3, 1, '2020-01-03 00:12:37', '13.4km', '20 mins', None),
 (4, 2, '2020-01-04 13:53:03', '23.4', '40', None),
 (5, 3, '2020-01-08 21:10:57', '10', '15', None),
 (6, 3, 'null', 'null', 'null', 'Restaurant Cancellation'),
 (7, 2, '2020-01-08 21:30:45', '25km', '25mins', 'null'),
 (8, 2, '2020-01-10 00:15:02', '23.4 km', '15 minute', 'null'),
 (9, 2, 'null', 'null', 'null', 'Customer Cancellation'),
 (10, 1, '2020-01-11 18:50:20', '10km', '10minutes', 'null')]

In [6]:
%%sql
select 
* 
from customer_orders;

 * sqlite:///pizza_runner.db
Done.


[(1, 101, 1, '', '', '2020-01-01 18:05:02'),
 (2, 101, 1, '', '', '2020-01-01 19:00:52'),
 (3, 102, 1, '', '', '2020-01-02 23:51:23'),
 (3, 102, 2, '', None, '2020-01-02 23:51:23'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 1, '4', '', '2020-01-04 13:23:46'),
 (4, 103, 2, '4', '', '2020-01-04 13:23:46'),
 (5, 104, 1, 'null', '1', '2020-01-08 21:00:29'),
 (6, 101, 2, 'null', 'null', '2020-01-08 21:03:13'),
 (7, 105, 2, 'null', '1', '2020-01-08 21:20:29'),
 (8, 102, 1, 'null', 'null', '2020-01-09 23:54:33'),
 (9, 103, 1, '4', '1, 5', '2020-01-10 11:22:59'),
 (10, 104, 1, 'null', 'null', '2020-01-11 18:34:49'),
 (10, 104, 1, '2, 6', '1, 4', '2020-01-11 18:34:49')]

In [7]:
%%sql
select 
* 
from pizza_names;

 * sqlite:///pizza_runner.db
Done.


[(1, 'Meat Lovers'), (2, 'Vegetarian')]

In [8]:
%%sql
select 
* 
from pizza_recipes;

 * sqlite:///pizza_runner.db
Done.


[(1, '1, 2, 3, 4, 5, 6, 8, 10'), (2, '4, 6, 7, 9, 11, 12')]

In [9]:
%%sql
select 
* 
from pizza_toppings;

 * sqlite:///pizza_runner.db
Done.


[(1, 'Bacon'),
 (2, 'BBQ Sauce'),
 (3, 'Beef'),
 (4, 'Cheese'),
 (5, 'Chicken'),
 (6, 'Mushrooms'),
 (7, 'Onions'),
 (8, 'Pepperoni'),
 (9, 'Peppers'),
 (10, 'Salami'),
 (11, 'Tomatoes'),
 (12, 'Tomato Sauce')]

### Case Study Questions

### A. Pizza Metrics Questions

#### Question 1 - How many pizzas were ordered?

In [10]:
%%sql
select count(*) ordered_pizzas
from customer_orders;

 * sqlite:///pizza_runner.db
Done.


[(14,)]

`14` pizzas were ordered.

#### Question 2 - How many unique customer orders were made?

In [11]:
%%sql
select count(distinct order_id) ordered_pizzas
from customer_orders;

 * sqlite:///pizza_runner.db
Done.


[(10,)]

`10` unique customer orders were made.

#### Question 3 - How many successful orders were delivered by each runner?

In [12]:
%%sql
select runner_id, count(order_id) successful_orders
from runner_orders
where duration not in ('null')
group by runner_id;

 * sqlite:///pizza_runner.db
Done.


[(1, 4), (2, 3), (3, 1)]

- Runner `1` delivered `4` orders successfully
- Runner `2` delivered `3` orders successfully
- Runner `3` had `1` successful order.

#### Question 4 - How many of each type of pizza was delivered?

In [13]:
%%sql
select co.pizza_id, pn.pizza_name, count(co.pizza_id) pizzas_delivered
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
join pizza_names pn
on co.pizza_id = pn.pizza_id
where ro.duration not in ('null')
group by co.pizza_id;

 * sqlite:///pizza_runner.db
Done.


[(1, 'Meat Lovers', 9), (2, 'Vegetarian', 3)]

- `Meatlovers` pizza was delivered `9` times
- `Vegetarian` pizza was delivered `3` times.

#### Question 5 - How many Vegetarian and Meatlovers were ordered by each customer?

In [14]:
%%sql
select co.customer_id, pn.pizza_name, count(co.pizza_id) pizzas_ordered
from customer_orders co
left join pizza_names pn
on co.pizza_id = pn.pizza_id
group by co.pizza_id,co.customer_id
order by co.customer_id;

 * sqlite:///pizza_runner.db
Done.


[(101, 'Meat Lovers', 2),
 (101, 'Vegetarian', 1),
 (102, 'Meat Lovers', 2),
 (102, 'Vegetarian', 1),
 (103, 'Meat Lovers', 3),
 (103, 'Vegetarian', 1),
 (104, 'Meat Lovers', 3),
 (105, 'Vegetarian', 1)]

- Customer `101` ordered `2` `MeatLovers` and `1` `Vegetarian`
- Customer `102` ordered `2` `MeatLovers` and `1` `Vegetarian`
- Customer `103` ordered `3` `Meatlovers` and `1` `Vegetarian`
- Customer `104` ordered `3` `Meatlovers`
- Customer `105` ordered `1` `Vegetarian`.

#### Question 6 - What was the maximum number of pizzas delivered in a single order?

In [15]:
%%sql
select ro.order_id, count(pizza_id)pizzas_delivered
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')
group by co.order_id
order by pizzas_delivered desc
limit 1;

 * sqlite:///pizza_runner.db
Done.


[(4, 3)]

- The maximum number of pizzas delivered in a single order is `3` which is order_id `4`.

#### Question 7 - For each customer, how many delivered pizzas had at least 1 change and how many had no changes?

In [16]:
%%sql
with ch as (
select co.customer_id
, case when co.exclusions or co.extras not in ('null', '') then 'change' 
    else 'no change' end change_flag
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')
order by co.customer_id
)
, chg as (
select *, count(c.change_flag) pizza_delvry_count
from ch c
group by  c.change_flag, c.customer_id
)

select *
from chg cg
order by cg.customer_id;

 * sqlite:///pizza_runner.db
Done.


[(101, 'no change', 2),
 (102, 'no change', 3),
 (103, 'change', 3),
 (104, 'change', 2),
 (104, 'no change', 1),
 (105, 'change', 1)]

- Customer `101` had `2` pizzas delivered with no change
- Customer `102` had `3` pizzas delivered with no change
- Customer `103` had `3` pizzas delivered with changes
- Customer `104` had `2` pizzas delivered with changes and `1` with no change
- Customer `105` had `1` pizza delivered with a change.

#### Question 8 - How many pizzas were delivered that had both exclusions and extras?

In [17]:
%%sql
select count(co.order_id) 'pizza count'
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')
and co.exclusions not in ('null', '')
and co.extras not in ('null', '')

 * sqlite:///pizza_runner.db
Done.


[(1,)]

- There was `1` pizza delivered that had both exclusions and extras.

#### Question 9 - What was the total volume of pizzas ordered for each hour of the day?

In [18]:
%%sql
select strftime('%H',datetime(co.order_time))'Hour', count(co.order_id) 'pizza volume'
from customer_orders co
group by strftime('%H',datetime(co.order_time))

 * sqlite:///pizza_runner.db
Done.


[('11', 1), ('13', 3), ('18', 3), ('19', 1), ('21', 3), ('23', 3)]

- `1` pizza was ordered at `11` hours
- `3` pizzas were ordered at `13` hours
- `3` pizzas were ordered at `18` hours
- `1` pizza was ordered at `19` hours
- `3` pizza were ordered at `21` hours
- `3` pizza were ordered at `23` hours.

#### Question 10 - What was the volume of orders for each day of the week?

In [19]:
%%sql
select strftime('%w',datetime(co.order_time))'Weekday', count(co.order_id) 'pizza volume'
from customer_orders co
group by strftime('%w',datetime(co.order_time))

 * sqlite:///pizza_runner.db
Done.


[('3', 5), ('4', 3), ('5', 1), ('6', 5)]

*** Having 0 == Sunday***
- `5` pizzas were ordered on `Wednesday`
- `3` pizzas were ordered on `Thursday`
- `1` pizza was ordered on `Friday`
- `5` pizzas were ordered on `Saturday`.

### B. Runner and Customer Experience Questions

#### Question 1 - How many runners signed up for each 1 week period? (i.e. week starts 2021-01-01)

In [20]:
%%sql
select r.registration_date
, count(strftime('%W',datetime(r.registration_date))) 'Runner Count'
from runners r
group by strftime('%W',datetime(r.registration_date));

 * sqlite:///pizza_runner.db
Done.


[('2021-01-01', 2), ('2021-01-08', 1), ('2021-01-15', 1)]

- In the first week, `2` runners registered and in the second & third week, `1` runner registered respectively.

#### Question 2 - What was the average time in minutes it took for each runner to arrive at the Pizza Runner HQ to pickup the order?

In [21]:
%%sql
select ro.runner_id, ro.pickup_time, co.order_time
, round(avg(cast((julianday(datetime(ro.pickup_time)) - 
        julianday(datetime(co.order_time))) * 24 * 60 As Integer)),0) average_arrival_time
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')
group by ro.runner_id

 * sqlite:///pizza_runner.db
Done.


[(1, '2020-01-01 18:15:34', '2020-01-01 18:05:02', 15.0),
 (2, '2020-01-04 13:53:03', '2020-01-04 13:23:46', 23.0),
 (3, '2020-01-08 21:10:57', '2020-01-08 21:00:29', 10.0)]

- On average it took runner 1 `15` minutes to arrive at the HQ for a pickup
- On average it took runner 2 `23` minutes to arrive at the HQ for a pickup
- On average it took runner 3 `10` minutes to arrive at the HQ for a pickup.

#### Question 3 - Is there any relationship between the number of pizzas and how long the order takes to prepare?

In [22]:
%%sql
select ro.runner_id, count(co.order_id) pizza_count
, round(avg(cast((julianday(datetime(ro.pickup_time)) - 
        julianday(datetime(co.order_time))) * 24 * 60 As Integer)),0) preparation_time
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')
group by ro.runner_id
order by preparation_time desc

 * sqlite:///pizza_runner.db
Done.


[(2, 5, 23.0), (1, 6, 15.0), (3, 1, 10.0)]

- There seems to be a positive relationship between the number of pizzas and the length of order preparation, i.e. the higher the number of pizzas, the more time it takes to prepare.

#### Question 4 - What was the average distance travelled for each customer?

In [23]:
%%sql
select co.customer_id, count(co.order_id)order_count
,round(avg(ro.distance),0)||'km' distance
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')
group by co.customer_id

 * sqlite:///pizza_runner.db
Done.


[(101, 2, '20.0km'),
 (102, 3, '17.0km'),
 (103, 3, '23.0km'),
 (104, 3, '10.0km'),
 (105, 1, '25.0km')]

- The average distance travelled for customer 101 is `20`km
- The average distance travelled for customer 102 is `17`km
- The average distance travelled for customer 103 is `23`km
- The average distance travelled for customer 104 is `10`km
- The average distance travelled for customer 105 is `25`km.

#### Question 5 - What was the difference between the longest and shortest delivery times for all orders?

In [24]:
%%sql
select max(ro.duration)||' minutes' longest_delivery
, REPLACE(min(ro.duration),'minutes', ' minutes') shortest_delivery
, (max(ro.duration) - min(ro.duration))||' minutes' difference
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
where ro.duration not in ('null')

 * sqlite:///pizza_runner.db
Done.


[('40 minutes', '10 minutes', '30 minutes')]

- The difference between the longest and shortest delivery times for all orders is `30`minutes.

#### Question 6 - What was the average speed for each runner for each delivery and do you notice any trend for these values?
speed = distance/duration

In [25]:
%%sql 
select ro.order_id, ro.runner_id, ro.distance, ro.duration
,round(avg((ro.distance*1.0 /ro.duration )*60),2)||'km/hr' average_speed
from runner_orders ro
where ro.duration not in ('null')
group by ro.runner_id, ro.order_id
order by ro.runner_id, order_id, average_speed 

 * sqlite:///pizza_runner.db
Done.


[(1, 1, '20km', '32 minutes', '37.5km/hr'),
 (2, 1, '20km', '27 minutes', '44.44km/hr'),
 (3, 1, '13.4km', '20 mins', '40.2km/hr'),
 (10, 1, '10km', '10minutes', '60.0km/hr'),
 (4, 2, '23.4', '40', '35.1km/hr'),
 (7, 2, '25km', '25mins', '60.0km/hr'),
 (8, 2, '23.4 km', '15 minute', '93.6km/hr'),
 (5, 3, '10', '15', '40.0km/hr')]

- On average, runner `2` seems to be the fastest amongst the runners 
- Additionally, there also tends to be an increase in the runners' average speed as their order delivery count increases.

#### Question 7 - What is the successful delivery percentage for each runner?

In [26]:
%%sql
with calc as (
select ro.runner_id, count(ro.order_id) tot_orders
,sum(case when ro.duration <> 'null' then 1 else 0 end) succ_delvry
from runner_orders ro
group by ro.runner_id
)

select c.runner_id
, cast((c.succ_delvry*1.0 /c.tot_orders)*100 as int)||'%' delv_percnt
from calc c

 * sqlite:///pizza_runner.db
Done.


[(1, '100%'), (2, '75%'), (3, '50%')]

- Runner `1` has a `100%` success rate
- Runner `2` has a `75%` success rate
- Runner `3` has a `50%` success rate.

### C. Ingredient Optimisation Questions

#### Question 1 - What are the standard ingredients for each pizza?

In [27]:
%%sql
/* transform toppings in the pizza recipes table into rows*/
with pizza_recipe as (
with recursive pr_split(pizza_id, toppings, str) as (
    select pizza_id, '', toppings||',' from pizza_recipes
    union all SELECT
    pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from pr_split where str!=''
) 
select pizza_id, toppings
from pr_split
where toppings!=''
order by pizza_id
)

select pr.pizza_id, pn.pizza_name, group_concat(pt.topping_name,', ') topping_name
from pizza_recipe pr
join pizza_names pn
on pr.pizza_id = pn.pizza_id
join pizza_toppings pt
on pr.toppings = pt.topping_id
group by pn.pizza_name;

 * sqlite:///pizza_runner.db
Done.


[(1, 'Meat Lovers', 'Bacon, BBQ Sauce, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (2, 'Vegetarian', 'Cheese, Mushrooms, Onions, Peppers, Tomatoes, Tomato Sauce')]

- Pizza 1's standard ingredients are `Bacon, BBQ Sauce, Beef, Cheese, Chicken, Mushrooms, Pepperoni & Salami`
- Pizza 2's standard ingredients are `Cheese, Mushrooms, Onions, Peppers, Tomatoes & Tomato Sauce`.

#### Question 2 - What was the most commonly added extra?

In [28]:
%%sql

/* transform extras into rows*/
with extras_transf as (
with recursive ex_split(pizza_id, extras, str) as (
select pizza_id, '', extras||',' from customer_orders
where extras not in ('null','')
union all 
select pizza_id,
substr(str, 0, instr(str, ',')),
substr(str, instr(str, ',')+1)
from ex_split 
where str!=''
) 
    
select pizza_id, extras
from ex_split
where extras!=''
order by pizza_id
    
)

/* transform toppings in the pizza recipes table into rows*/
, pizza_recipe as (
with recursive pr_split(pizza_id, toppings, str) as (
select pizza_id, '', toppings||',' from pizza_recipes
union all 
select pizza_id,
substr(str, 0, instr(str, ',')),
substr(str, instr(str, ',')+1)
from pr_split 
where str!=''
) 
    
select pizza_id, toppings
from pr_split
where toppings!=''
order by pizza_id
    
)

select pt.topping_name, count(et.extras) count_of_extras
from extras_transf et
join pizza_toppings pt
on et.extras = pt.topping_id
group by pt.topping_name
order by pt.topping_name

 * sqlite:///pizza_runner.db
Done.


[('Bacon', 4), ('Cheese', 1), ('Chicken', 1)]

- As can be seen, the most commonly added extra was `Bacon`.

#### Question 3 - What was the most common exclusion?

In [29]:
%%sql

/* transform exclusions into rows*/
with exclusions_transf as (
with recursive exc_split(pizza_id, exclusions, str) as (
select pizza_id, '', exclusions||',' from customer_orders
where exclusions not in ('null','')
union all 
select pizza_id,
substr(str, 0, instr(str, ',')),
substr(str, instr(str, ',')+1)
from exc_split 
where str!=''
) 
    
select pizza_id, exclusions
from exc_split
where exclusions!=''
order by pizza_id
    
)

/* transform toppings in the pizza recipes table into rows*/
, pizza_recipe as (
with recursive pr_split(pizza_id, toppings, str) as (
select pizza_id, '', toppings||',' from pizza_recipes
union all 
select pizza_id,
substr(str, 0, instr(str, ',')),
substr(str, instr(str, ',')+1)
from pr_split 
where str!=''
) 
    
select pizza_id, toppings
from pr_split
where toppings!=''
order by pizza_id
    
)

select pt.topping_name, count(et.exclusions) count_of_exclusions
from exclusions_transf et
join pizza_toppings pt
on et.exclusions = pt.topping_id
group by pt.topping_name
order by et.exclusions desc

 * sqlite:///pizza_runner.db
Done.


[('Cheese', 4), ('BBQ Sauce', 1), ('Mushrooms', 1)]

- As can be seen, the most common exclusion was `Cheese`.

#### Question 4 - Generate an order item for each record in the customers_orders table in the format of one of the following:
- Meat Lovers
- Meat Lovers - Exclude Beef
- Meat Lovers - Extra Bacon
- Meat Lovers - Exclude Cheese, Bacon - Extra Mushroom, Peppers

In [30]:
%%sql
with compiled as (
with leng as (
select order_id, customer_id, pizza_id, exclusions, extras
,replace(replace(exclusions,'null',''),', ','') clean
,length(replace(replace(exclusions,'null',''),', ','')) length_excl
,length(replace(replace(extras,'null',''),', ','')) length_ext
from customer_orders 
)
    
select order_id, customer_id, pizza_id, exclusions
,extras, length_excl, length_ext
, case when  length_excl > 1 and length_ext > 1 then 
(/*exclusion split */
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng 
    where length_excl  > 1 and length_ext  > 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select group_concat(pt.topping_name, ', ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
group by es.pizza_id
) 

when  length_excl  > 1 and length_ext  == 1 then 
(/*exclusion split*/
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng 
    where length_excl  > 1 and length_ext  == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
group by es.pizza_id
) 

when length_excl  == 1 then 
(/*exclusion split*/
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng where length_excl  == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select pt.topping_name
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
--group by es.pizza_id
--and exclusions not in ('null')
--order by pizza_id
) 
else '' end excl_flag


, case when  length_ext  > 1 and length_excl > 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  > 1 and length_excl > 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ')
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where es1.extras!=''
group by es1.pizza_id
) 

when  length_ext  > 1 and length_excl == 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  > 1 and length_excl == 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ') 
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where es1.extras!=''
group by es1.pizza_id
) 

when length_ext  == 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  == 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select pt.topping_name
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where extras!=''
) 
else '' end ext_flag

from leng 
)

select c.order_id
, case when exclusions not in ('null','') and extras not in ('null','') then
pn.pizza_name||' - Exclude '||excl_flag||' - Extra '||ext_flag
when exclusions not in ('null','') and extras in ('null','') then
pn.pizza_name||' - Exclude '||excl_flag
when exclusions in ('null','') and extras not in ('null','') then
pn.pizza_name||' - Extra '||ext_flag
else pn.pizza_name end ordered_items
from compiled c
join pizza_names pn
on c.pizza_id = pn.pizza_id

 * sqlite:///pizza_runner.db
Done.


[(1, 'Meat Lovers'),
 (2, 'Meat Lovers'),
 (3, 'Meat Lovers'),
 (3, 'Vegetarian'),
 (4, 'Meat Lovers - Exclude Cheese'),
 (4, 'Meat Lovers - Exclude Cheese'),
 (4, 'Vegetarian - Exclude Cheese'),
 (5, 'Meat Lovers - Extra Bacon'),
 (6, 'Vegetarian'),
 (7, 'Vegetarian - Extra Bacon'),
 (8, 'Meat Lovers'),
 (9, 'Meat Lovers - Exclude Cheese - Extra Bacon, Chicken'),
 (10, 'Meat Lovers'),
 (10, 'Meat Lovers - Exclude BBQ Sauce, Mushrooms - Extra Bacon, Cheese')]

#### Question 5 - Generate an alphabetically ordered comma separated ingredient list for each pizza order from the `customer_orders` table and add a `2x` in front of any relevant ingredients
- For example: "Meat Lovers: 2xBacon, Beef, ... , Salami"

In [31]:
%%sql
with newfull2 as (
with newfull1 as  (
with newfull as (
with full as (
with compiled as (
with leng as (
select order_id, customer_id, pizza_id, exclusions, extras
,replace(replace(exclusions,'null',''),', ','') clean
,length(replace(replace(exclusions,'null',''),', ','')) length_excl
,length(replace(replace(extras,'null',''),', ','')) length_ext
from customer_orders 
)
    
select order_id, customer_id, pizza_id, exclusions
,extras, length_excl, length_ext
, case when  length_excl > 1 and length_ext > 1 then 
(/*exclusion split */
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng 
    where length_excl  > 1 and length_ext  > 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select group_concat(pt.topping_name, ', ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
group by es.pizza_id
) 

when  length_excl  > 1 and length_ext  == 1 then 
(/*exclusion split*/
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng 
    where length_excl  > 1 and length_ext  == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
group by es.pizza_id
) 

when length_excl  == 1 then 
(/*exclusion split*/
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng where length_excl  == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select trim(pt.topping_name,' ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
--group by es.pizza_id
--and exclusions not in ('null')
--order by pizza_id
) 
else '' end excl_flag

, case when  length_ext  > 1 and length_excl > 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  > 1 and length_excl > 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ')
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where es1.extras!=''
group by es1.pizza_id
) 

when  length_ext  > 1 and length_excl == 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  > 1 and length_excl == 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ') 
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where es1.extras!=''
group by es1.pizza_id
) 

when length_ext  == 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  == 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select pt.topping_name
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where extras!=''
) 
else '' end ext_flag
from leng 
)

, comb_rec as (
/* transform toppings in the pizza recipes table into rows*/
with pizza_recipe as (
with recursive pr_split(pizza_id, toppings, str) as (
    select pizza_id, '', toppings||',' from pizza_recipes
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from pr_split where str!=''
) 
select pizza_id, toppings
from pr_split
where toppings!=''
order by pizza_id
)

select pr.pizza_id, pn.pizza_name, group_concat(pt.topping_name,', ') topping_name
from pizza_recipe pr
join pizza_names pn
on pr.pizza_id = pn.pizza_id
join pizza_toppings pt
on pr.toppings = pt.topping_id
group by pn.pizza_name
)
    
select c.order_id, c.customer_id, c.pizza_id, c.exclusions, c.extras
, c.length_excl, c.length_ext, c.excl_flag, c.ext_flag, pn.pizza_name, cr.topping_name
, case when exclusions not in ('null','') and extras not in ('null','') then
pn.pizza_name||' - Exclude '||excl_flag||' - Extra '||ext_flag
when exclusions not in ('null','') and extras in ('null','') then
pn.pizza_name||' - Exclude '||excl_flag
when exclusions in ('null','') and extras not in ('null','') then
pn.pizza_name||' - Extra '||ext_flag
else pn.pizza_name end ordered_items
from compiled c
join pizza_names pn
on c.pizza_id = pn.pizza_id
join comb_rec cr
on c.pizza_id = cr.pizza_id

)

select *
, case when length_excl>1 and length_ext>1 then
(select topping_name||', '||excl_flag from full where length_excl>1 and length_ext>1)
when length_excl==1 and length_ext>1 then 
(select topping_name||', '||excl_flag from full where length_excl==1 and length_ext>1)
when length_excl==1 and length_ext<1 and pizza_id ==1 then 
(select trim(topping_name,' ')||', '||trim(excl_flag,' ') from full where length_excl==1 and length_ext<1 and pizza_id==1)
when length_excl==1 and length_ext<1 and pizza_id ==2 then 
(select trim(topping_name,' ')||', '||trim(excl_flag,' ') from full where length_excl==1 and length_ext<1 and pizza_id==2)    
else topping_name end excl_merg

, case when length_ext>1 and length_excl>1 then 
(select topping_name||', '||ext_flag from full where length_ext>1 and length_excl>1)
when length_ext>1 and length_excl==1 then 
(select topping_name||', '||ext_flag from full where length_ext>1 and length_excl==1)
when length_ext==1 and length_excl<1 and pizza_id==1 then 
(select topping_name||', '||ext_flag from full where length_ext==1 and length_excl<1 and pizza_id==1)
when length_ext==1 and length_excl<1 and pizza_id==2 then 
(select topping_name||', '||ext_flag from full where length_ext==1 and length_excl<1 and pizza_id==2)
else topping_name end ext_merg

from full

)

select  *
, case when length_excl>1 and length_ext>1 then
(
/*exclusion_merge split */
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl>1 and length_ext>1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id
having count <2 

)
   select group_concat(excl_merg,',')
    from splitmerg
    group by pizza_id
)

  
when length_excl==1 and length_ext>1 then 
(
/*exclusion_merge split */
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl==1 and length_ext>1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id
having count <2 

)
   select group_concat(excl_merg,',')
    from splitmerg
    group by pizza_id
)
        
when length_excl==1 and length_ext<1 and pizza_id==1 then 
(
/*exclusion_merge split */
with splitmerg1 as(
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl==1 and length_ext<1 and pizza_id ==1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id

)
   select *, (count/2 /*2 is the count of items where length_excl==1 and length<1 and pizza_id==1*/) counts 
    from splitmerg
    group by excl_merg
    having counts <2
)
   select group_concat(trim(excl_merg,' '),',')
    from splitmerg1
    group by pizza_id
)

when length_excl==1 and length_ext<1 and pizza_id==2 then
(
/*exclusion_merge split */
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl==1 and length_ext<1 and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id
having count <2 

)
   select group_concat(excl_merg,',')
    from splitmerg
    group by pizza_id
)    
else excl_merg end excl_merg1
    
from newfull

)
select *, (replace(excl_merg1,', ',',')||','||replace(ext_merg,', ',',')) combinedexcl_extmerg
from newfull1
    
)

select  order_id, 
case when length_excl>1 and length_ext>1 then 
(/*combinedexcl_extmerg split */
    with strip_close as (   
    with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl > 1 and length_ext>1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select  
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 3 then '2x'||combinedexcl_extmerg
    when count == 2 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg
)

     select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close
    
)

when length_excl == 1 and length_ext > 1 then 
(/*combinedexcl_extmerg split */
    with strip_close as (  
    with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl == 1 and length_ext>1 
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)

    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 3 then '2x'||combinedexcl_extmerg
    when count == 2 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close

)

when length_excl == 1 and length_ext < 1 and pizza_id==1 then 
(/*combinedexcl_extmerg split */
    with strip_close as (   
    with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', combinedexcl_extmerg||',' from newfull2 
    where length_excl == 1 and length_ext < 1 and pizza_id==1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)  
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
  , case when count > 2 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close
    group by pizza_id
)

when length_excl == 1 and length_ext < 1 and pizza_id==2 then 
(/*combinedexcl_extmerg split */
    with strip_close as (   
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl == 1 and length_ext<1 and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
) 
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 2 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close
)

when length_excl < 1 and ifnull(length_ext,0) < 1  and pizza_id==2 then 
(/*combinedexcl_extmerg split */
    with strip_close as (    
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and ifnull(length_ext,0) <1  and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 4 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close
)

when length_excl < 1 and length_ext == 1  and pizza_id==1 then 
(/*combinedexcl_extmerg split */
    with strip_close as (    
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext==1  and pizza_id==1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 3 then '2x'||combinedexcl_extmerg 
    when count == 2 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close
    
)

when length_excl < 1 and length_ext == 1  and pizza_id==2 then 
(/*combinedexcl_extmerg split */
    with strip_close as ( 
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext==1  and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)    
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count <= 2 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close
    
)

when length_excl < 1 and length_ext<1 and pizza_id==1 then 
(/*combinedexcl_extmerg split */
    with strip_close as (
    
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext<1  and pizza_id == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 10 then combinedexcl_extmerg else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select pizza_name||': '||group_concat(order_list,', ')order_list
    from strip_close    
)
else '' end order_list

from newfull2;

 * sqlite:///pizza_runner.db
Done.


[(1, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (2, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (3, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (3, 'Vegetarian: Cheese, Mushrooms, Onions, Peppers, Tomato Sauce, Tomatoes'),
 (4, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Chicken, Mushrooms, Pepperoni, Salami'),
 (4, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Chicken, Mushrooms, Pepperoni, Salami'),
 (4, 'Vegetarian: Mushrooms, Onions, Peppers, Tomato Sauce, Tomatoes'),
 (5, 'Meat Lovers: BBQ Sauce, 2xBacon, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (6, 'Vegetarian: Cheese, Mushrooms, Onions, Peppers, Tomato Sauce, Tomatoes'),
 (7, 'Vegetarian: Bacon, Cheese, Mushrooms, Onions, Peppers, Tomato Sauce, Tomatoes'),
 (8, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (9, 'Meat Lovers: BBQ Sauce, 2xBacon, Beef, 2xChicken, Mushrooms, Pepperoni, Salami'),
 (10, 'Meat Lovers: BBQ Sauce, Bacon, Beef, Cheese, Chicken, Mushrooms, Pepperoni, Salami'),
 (10, 'Meat Lovers: 2xBacon, Beef, 2xCheese, Chicken, Pepperoni, Salami')]

#### Question 6 - What is the total quantity of each ingredient used in all delivered pizzas sorted by most frequent first?

In [32]:
%%sql
with newfull2 as (
with newfull1 as  (
with newfull as (
with full as (
with compiled as (
with leng as (
select co.order_id, customer_id, pizza_id, exclusions, extras
,replace(replace(exclusions,'null',''),', ','') clean
,length(replace(replace(exclusions,'null',''),', ','')) length_excl
,length(replace(replace(ifnull(extras,'null'),'null',''),', ','')) length_ext
from runner_orders ro
join customer_orders co
on ro.order_id == co.order_id
where ro.distance not in ('null')
)
    
select order_id, customer_id, pizza_id, exclusions
,extras, length_excl, length_ext
, case when  length_excl > 1 and length_ext > 1 then 
(/*exclusion split */
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng 
    where length_excl  > 1 and length_ext  > 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select group_concat(pt.topping_name, ', ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
group by es.pizza_id
) 

when  length_excl  > 1 and length_ext  == 1 then 
(/*exclusion split*/
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng 
    where length_excl  > 1 and length_ext  == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
group by es.pizza_id
) 

when length_excl  == 1 then 
(/*exclusion split*/
with recursive excl_split(pizza_id, exclusions, str) as (
    select pizza_id, '', exclusions||',' from leng where length_excl  == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from excl_split 
    where str!=''
    
) 
select trim(pt.topping_name,' ')
from excl_split es
left join pizza_toppings pt
on es.exclusions = pt.topping_id
where exclusions!=''
--group by es.pizza_id
--and exclusions not in ('null')
--order by pizza_id
) 
else '' end excl_flag

, case when  length_ext  > 1 and length_excl > 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  > 1 and length_excl > 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ')
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where es1.extras!=''
group by es1.pizza_id
) 

when  length_ext  > 1 and length_excl == 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  > 1 and length_excl == 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select group_concat(pt.topping_name,', ') 
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where es1.extras!=''
group by es1.pizza_id
) 

when length_ext  == 1 then 
(/*extras split*/
with recursive ext_split(pizza_id, extras, str) as (
    select pizza_id, '', extras||',' from leng 
    where length_ext  == 1
    union all
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from ext_split 
    where str!=''
    
) 
select pt.topping_name
from ext_split es1
left join pizza_toppings pt
on es1.extras = pt.topping_id
where extras!=''
) 
else '' end ext_flag
from leng 
)

, comb_rec as (
/* transform toppings in the pizza recipes table into rows*/
with pizza_recipe as (
with recursive pr_split(pizza_id, toppings, str) as (
    select pizza_id, '', toppings||',' from pizza_recipes
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from pr_split where str!=''
) 
select pizza_id, toppings
from pr_split
where toppings!=''
order by pizza_id
)

select pr.pizza_id, pn.pizza_name, group_concat(pt.topping_name,', ') topping_name
from pizza_recipe pr
join pizza_names pn
on pr.pizza_id = pn.pizza_id
join pizza_toppings pt
on pr.toppings = pt.topping_id
group by pn.pizza_name
)
    
select c.order_id, c.customer_id, c.pizza_id, c.exclusions, c.extras
, c.length_excl, c.length_ext, c.excl_flag, c.ext_flag, pn.pizza_name, cr.topping_name
, case when exclusions not in ('null','') and extras not in ('null','') then
pn.pizza_name||' - Exclude '||excl_flag||' - Extra '||ext_flag
when exclusions not in ('null','') and extras in ('null','') then
pn.pizza_name||' - Exclude '||excl_flag
when exclusions in ('null','') and extras not in ('null','') then
pn.pizza_name||' - Extra '||ext_flag
else pn.pizza_name end ordered_items
from compiled c
join pizza_names pn
on c.pizza_id = pn.pizza_id
join comb_rec cr
on c.pizza_id = cr.pizza_id

)

select *
, case when length_excl>1 and length_ext>1 then
(select topping_name||', '||excl_flag from full where length_excl>1 and length_ext>1)
when length_excl==1 and length_ext>1 then 
(select topping_name||', '||excl_flag from full where length_excl==1 and length_ext>1)
when length_excl==1 and length_ext<1 and pizza_id ==1 then 
(select trim(topping_name,' ')||', '||trim(excl_flag,' ') from full where length_excl==1 and length_ext<1 and pizza_id==1)
when length_excl==1 and length_ext<1 and pizza_id ==2 then 
(select trim(topping_name,' ')||', '||trim(excl_flag,' ') from full where length_excl==1 and length_ext<1 and pizza_id==2)    
else topping_name end excl_merg

, case when length_ext>1 and length_excl>1 then 
(select topping_name||', '||ext_flag from full where length_ext>1 and length_excl>1)
when length_ext>1 and length_excl==1 then 
(select topping_name||', '||ext_flag from full where length_ext>1 and length_excl==1)
when length_ext==1 and length_excl<1 and pizza_id==1 then 
(select topping_name||', '||ext_flag from full where length_ext==1 and length_excl<1 and pizza_id==1)
when length_ext==1 and length_excl<1 and pizza_id==2 then 
(select topping_name||', '||ext_flag from full where length_ext==1 and length_excl<1 and pizza_id==2)
else topping_name end ext_merg

from full

)

select  *
, case when length_excl>1 and length_ext>1 then
(
/*exclusion_merge split */
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl>1 and length_ext>1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id
having count <2 

)
   select group_concat(excl_merg,',')
    from splitmerg
    group by pizza_id
)

  
when length_excl==1 and length_ext>1 then 
(
/*exclusion_merge split */
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl==1 and length_ext>1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id
having count <2 

)
   select group_concat(excl_merg,',')
    from splitmerg
    group by pizza_id
)
        
when length_excl==1 and length_ext<1 and pizza_id==1 then 
(
/*exclusion_merge split */
with splitmerg1 as(
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl==1 and length_ext<1 and pizza_id ==1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id

)
   select *, (count/2 /*2 is the count of items where length_excl==1 and length<1 and pizza_id==1*/) counts 
    from splitmerg
    group by excl_merg
    having counts <2
)
   select group_concat(trim(excl_merg,' '),',')
    from splitmerg1
    group by pizza_id
)

when length_excl==1 and length_ext<1 and pizza_id==2 then
(
/*exclusion_merge split */
with splitmerg as  (
    with recursive exclmerg_split(pizza_id, excl_merg, str) as (
    select pizza_id, '', excl_merg||',' from newfull 
    where length_excl==1 and length_ext<1 and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
select pizza_id, es2.excl_merg, count(es2.excl_merg) count  
from exclmerg_split es2
where excl_merg!=''
group by es2.excl_merg --es.pizza_id
having count <2 

)
   select group_concat(excl_merg,',')
    from splitmerg
    group by pizza_id
)    
else excl_merg end excl_merg1
    
from newfull

)
select *, (replace(excl_merg1,', ',',')||','||replace(ext_merg,', ',',')) combinedexcl_extmerg
from newfull1
    
)


, exg1etg1 as (
with strip_close as ( 
        /*combinedexcl_extmerg split */
    with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl > 1 and length_ext>1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select  
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 3 then 2
    when count == 2 then 1 else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg
)

     select combinedexcl_extmerg, order_list
     from strip_close
)        

, exe1etg1 as (
with strip_close as ( /*combinedexcl_extmerg split */ 
    with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl == 1 and length_ext>1 
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)

    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 3 then 2
    when count == 2 then 1 else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list
    from strip_close
    )  

, exe1etl1ide1 as (
with strip_close as ( 
        /*combinedexcl_extmerg split */  
    with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', combinedexcl_extmerg||',' from newfull2 
    where length_excl == 1 and length_ext < 1 and pizza_id==1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)  
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
  , case when count > 2 then (select count(*) from newfull2 where length_excl == 1 and length_ext < 1 and pizza_id==1)
    else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list 
    from strip_close
    )    

, exe1etl1ide2 as (
with strip_close as ( 
        /*combinedexcl_extmerg split */
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl == 1 and length_ext<1 and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
) 
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 2 then (select count(*) from newfull2 where length_excl == 1 and length_ext < 1 and pizza_id==2)  
    else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list 
    from strip_close
    ) 

, exl1etl1ide2 as (
with strip_close as (   
        /*combinedexcl_extmerg split */
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext <1 and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count >= 2 then (select count(*) from newfull2 where length_excl < 1 and length_ext <1 and pizza_id==2)   
    else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list 
    from strip_close
    )  

, exl1ete1ide1 as (
with strip_close as (   
    /*combinedexcl_extmerg split */
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext==1  and pizza_id==1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 3 then 2 
    when count == 2 then 1 else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list 
    from strip_close
    )
    
, exl1ete1ide2 as (
with strip_close as ( 
    /*combinedexcl_extmerg split */
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext==1  and pizza_id==2
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)    
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count <= 2 then (select count(*) from newfull2 where length_excl<1 and length_ext==1 and pizza_id==2) 
    else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list 
    from strip_close
    )
     
, exl1etl1ide1 as (
with strip_close as (
    /*combinedexcl_extmerg split */
with stripped as (
    with recursive exclmerg_split(pizza_id, combinedexcl_extmerg, str) as (
    select pizza_id, '', trim(combinedexcl_extmerg,' ')||',' from newfull2 
    where length_excl < 1 and length_ext<1  and pizza_id == 1
    union all 
    select pizza_id,
    substr(str, 0, instr(str, ',')),
    substr(str, instr(str, ',')+1)
    from exclmerg_split 
    where str!=''
    
) 
    select trim(es2.combinedexcl_extmerg,' ')combinedexcl_extmerg , es2.pizza_id
    , count(es2.combinedexcl_extmerg) over(partition by es2.combinedexcl_extmerg) count
    from exclmerg_split es2
    where combinedexcl_extmerg!=''
    order by es2.combinedexcl_extmerg asc
)
    select 
    distinct trim(combinedexcl_extmerg,' ')combinedexcl_extmerg,pizza_id,count 
    , case when count == 10 then (select count(*) from newfull2 where length_excl < 1 and length_ext<1  and pizza_id == 1)
    else '' end order_list
    from stripped where order_list!=''
    order by combinedexcl_extmerg 
)
    select combinedexcl_extmerg, order_list 
    from strip_close    
    )
    
, tot_sum as (
    select combinedexcl_extmerg ingredients, order_list from exg1etg1
    union all
    select combinedexcl_extmerg, order_list from exe1etg1
    union all
    select combinedexcl_extmerg, order_list from exe1etl1ide1
    union all
    select combinedexcl_extmerg, order_list from exe1etl1ide2
    union all
    select combinedexcl_extmerg, order_list from exl1etl1ide2
    union all
    select combinedexcl_extmerg, order_list from exl1ete1ide1
    union all
    select combinedexcl_extmerg, order_list from exl1ete1ide2
    union all
    select combinedexcl_extmerg, order_list from exl1etl1ide1
)

select ingredients, sum(order_list) total_ingredient
from tot_sum
group by ingredients
order by total_ingredient desc;

 * sqlite:///pizza_runner.db
Done.


[('Bacon', 12),
 ('Mushrooms', 11),
 ('Cheese', 10),
 ('Salami', 9),
 ('Pepperoni', 9),
 ('Chicken', 9),
 ('Beef', 9),
 ('BBQ Sauce', 8),
 ('Tomatoes', 3),
 ('Tomato Sauce', 3),
 ('Peppers', 3),
 ('Onions', 3)]

### D. Pricing and Ratings Questions

#### Question 1 - If a Meat Lovers pizza costs `$12` and Vegetarian costs `$10` and there were no charges for changes - how much money has Pizza Runner made so far if there are no delivery fees?

In [33]:
%%sql
select pn.pizza_name, '$'||sum(case when co.pizza_id = 1 then 12 else 10 end) sales
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
join pizza_names pn
on co.pizza_id = pn.pizza_id
where ro.duration not in ('null')
group by co.pizza_id

 * sqlite:///pizza_runner.db
Done.


[('Meat Lovers', '$108'), ('Vegetarian', '$30')]

- Pizza Runner would have made `$108` on `Meat Lovers` pizza and `$30` on `Vegetarian` pizza
- In total Pizza Runner would have made `$138`.

#### Question 2 - What if there was an additional `$1` charge for any pizza extras?
- Add cheese is `$1` extra

In [34]:
%%sql
with add_char as (
select pn.pizza_name
, case when co.extras not in ('null','') then length(replace(co.extras,', ','')) 
    else 0 end extras_count
,case when co.pizza_id = 1 then 12 else 10 end price
from customer_orders co
left join runner_orders ro
on co.order_id = ro.order_id
join pizza_names pn
on co.pizza_id = pn.pizza_id
where ro.duration not in ('null')
)

select pizza_name, '$'||sum((extras_count + price)) revenue
from add_char
group by pizza_name

 * sqlite:///pizza_runner.db
Done.


[('Meat Lovers', '$111'), ('Vegetarian', '$31')]

- Pizza Runner would have made `$111` on `Meat Lovers` pizza and `$31` on `Vegetarian` pizza
- In total Pizza Runner would have made `$142`.

#### Question 3 - The Pizza Runner team now wants to add an additional ratings system that allows customers to rate their runner, how would you design an additional table for this new dataset - generate a schema for this new table and insert your own data for ratings for each successful customer order between 1 to 5.
- Table has been created above with the other tables 

In [35]:
%%sql
select *
from ratings;

 * sqlite:///pizza_runner.db
Done.


[(1, 5, '2020-01-01 17:20:00'),
 (2, 5, '2020-01-01 18:30:34'),
 (3, 4, '2020-01-03 01:10:47'),
 (4, 3, '2020-01-04 15:33:33'),
 (5, 3, '2020-01-08 23:20:31'),
 (6, 'null', 'null'),
 (7, 3, '2020-01-09 20:30:45'),
 (8, 4, '2020-01-10 01:15:02'),
 (9, 'null', 'null'),
 (10, 2, '2020-01-12 12:50:20')]

- The `ratings` table was created to hold order_id, ratings and time of ratings
- With the `order_id` it can be joined with the customer_orders and runner_orders table to extract other information.

#### Question 4 - Using your newly generated table - can you join all of the information together to form a table which has the following information for successful deliveries?
- customer_id
- order_id
- runner_id
- rating
- order_time
- pickup_time
- Time between order and pickup
- Delivery duration
- Average speed
- Total number of pizzas

In [36]:
%%sql
select co.customer_id, co.order_id, ro.runner_id, r.ratings, co.order_time, ro.pickup_time
, round(cast((julianday(datetime(ro.pickup_time)) - 
        julianday(datetime(co.order_time))) * 24 * 60 As Integer),0) arrival_time
, ro.duration,round(avg((ro.distance*1.0 /ro.duration*60)),2)||'km/hr' average_speed
, count(ro.order_id) pizza_count
from runner_orders ro
join customer_orders co
on co.order_id = ro.order_id
join ratings r
on ro.order_id = r.order_id
where ro.duration not in ('null')
group by co.customer_id, co.order_id, ro.runner_id, r.ratings
, co.order_time, ro.pickup_time

 * sqlite:///pizza_runner.db
Done.


[(101, 1, 1, 5, '2020-01-01 18:05:02', '2020-01-01 18:15:34', 10.0, '32 minutes', '37.5km/hr', 1),
 (101, 2, 1, 5, '2020-01-01 19:00:52', '2020-01-01 19:10:54', 10.0, '27 minutes', '44.44km/hr', 1),
 (102, 3, 1, 4, '2020-01-02 23:51:23', '2020-01-03 00:12:37', 21.0, '20 mins', '40.2km/hr', 2),
 (102, 8, 2, 4, '2020-01-09 23:54:33', '2020-01-10 00:15:02', 20.0, '15 minute', '93.6km/hr', 1),
 (103, 4, 2, 3, '2020-01-04 13:23:46', '2020-01-04 13:53:03', 29.0, '40', '35.1km/hr', 3),
 (104, 5, 3, 3, '2020-01-08 21:00:29', '2020-01-08 21:10:57', 10.0, '15', '40.0km/hr', 1),
 (104, 10, 1, 2, '2020-01-11 18:34:49', '2020-01-11 18:50:20', 15.0, '10minutes', '60.0km/hr', 2),
 (105, 7, 2, 3, '2020-01-08 21:20:29', '2020-01-08 21:30:45', 10.0, '25mins', '60.0km/hr', 1)]

#### Question 5 - If a Meat Lovers pizza was `$12` and Vegetarian `$10` fixed prices with no cost for extras and each runner is paid $0.30 per kilometre traveled - how much money does Pizza Runner have left over after these deliveries?


In [37]:
%%sql
select '$'||sum(round((sales - delvry_cost),2)) net_profit
from
(select ro.order_id, ro.runner_id, ro.distance
, sum(case when co.pizza_id = 1 then 12 else 10 end) sales
, ro.distance*0.30 delvry_cost
from runner_orders ro
left join customer_orders co
on co.order_id = ro.order_id
where ro.duration not in ('null')
group by ro.order_id
order by ro.order_id, ro.runner_id, sales
) calc 

 * sqlite:///pizza_runner.db
Done.


[('$94.44',)]

- After deliverires, Pizza Runner have a net profit of `$94.44`.